File: 2976-minimum-cost-to-convert-string-i/2976-minimum-cost-to-convert-string-i.cpp
class Solution {
public:
    long long minimumCost(string source, string target, vector<char>& original, vector<char>& changed, vector<int>& cost) {
        long long res = 0;
        unordered_map<char, unordered_map<char, int>> graph;
        unordered_map<char, unordered_map<char, long long>> minCost;
        for(int i = 0; i < original.size(); i++)
        {
            if(graph[original[i]].count(changed[i]))
            {
                graph[original[i]][changed[i]] = min(graph[original[i]][changed[i]], cost[i]);
            }
            else
            {
                graph[original[i]][changed[i]] = cost[i];
            }
        }
        for(char i = 'a'; i <= 'z'; i++)
        {
            minCost[i] = dijk(i, graph);
        }
        for(int i = 0; i < source.length(); i++)
        {
            if(minCost[source[i]].count(target[i]))
            {
                res += minCost[source[i]][target[i]];
            }
            else
            {
                return -1;
            }
        }
        return res;
    }
private:
    unordered_map<char, long long> dijk(char source, unordered_map<char, unordered_map<char, int>>& graph)
    {
        unordered_map<char, long long> res;
        priority_queue<pair<long long, char>, vector<pair<long long, char>>, greater<pair<long long, char>>> pq;
        unordered_set<char> visited;
        res[source] = 0;
        pq.push({0, source});
        while(!pq.empty())
        {
            char dest = pq.top().second;
            long long cost = pq.top().first;
            pq.pop();
            if(visited.count(dest))
            {
                continue;
            }
            visited.insert(dest);
            for(auto it : graph[dest])
            {
                if(!res.count(it.first))
                {
                    res[it.first] = it.second + cost;
                }
                else
                {
                    res[it.first] = min(res[it.first], it.second + cost);
                }
                pq.push({it.second + cost, it.first});
            }
        }
        return res;
    }
};

File: 2976-minimum-cost-to-convert-string-i/2976-minimum-cost-to-convert-string-i.py
import heapq
from collections import defaultdict
class Solution:
    def minimumCost(self, source: str, target: str, original: list[str], changed: list[str], cost: list[int]) -> int:
        graph = defaultdict(lambda: defaultdict(lambda: float('inf')))
        for o, c, z in zip(original, changed, cost):
            graph[o][c] = min(graph[o][c], z)
        min_costs = {}
        def dijkstra(start_node):
            distances = {start_node: 0}
            pq = [(0, start_node)]
            while pq:
                curr_dist, u = heapq.heappop(pq)
                if curr_dist > distances.get(u, float('inf')):
                    continue
                for v, weight in graph[u].items():
                    if curr_dist + weight < distances.get(v, float('inf')):
                        distances[v] = curr_dist + weight
                        heapq.heappush(pq, (distances[v], v))
            return distances
        unique_chars = set(source)
        for char in unique_chars:
            min_costs[char] = dijkstra(char)
        total_cost = 0
        for s, t in zip(source, target):
            if s == t:
                continue
            if t in min_costs[s]:
                total_cost += min_costs[s][t]
            else:
                return -1    
        return total_cost